Improve JavaScript examples and add chapter for npm and ESM (#31)
rism-digital/verovio-reference-book: _book/02-first-steps/05-npm-usage.md

---

> title: "Usage with npm"

Reviewer: How would one decide whether they want to use the node, esm, or cjs methods? There should probably be an introduction to what this is, and what people can expect to find on this page. Right now it's quite detailed in specifics, but with no introduction it makes it hard to understand whether I (as a random reader) need to pay attention to this page.

Reviewer: Also, there's no info on this page on how you might include it in a bundler, like Webpack. How would you install verovio through NPM and then configure it to build a bundle, and how would you include that bundle in an HTML page?

Author: My repository with the examples is not migrated yet for the final implementation of my pull request https://github.com/rism-digital/verovio/pull/2946, but it's really similar.

Reviewer 2: > I could add them to the verovio repository in ~/emscripten/npm/examples and link to the code examples in this tutorial, what do you think? @lpugin Is there a better place for the examples?

Yes, I think this would be a reasonable place for having them.

---

> The latest stable version is available via NPM registry. The version distributed via NPM it the WebAssembly build. It can be installed with: 

Reviewer: ```suggestion
The latest stable version is available in the NPM registry. The version distributed via NPM is the WebAssembly build. It can be installed with: 
```

```bash
npm install verovio
```

> ## Basic usage

Reviewer: This is basic usage for what? Node environments? Given that the vast majority of people will use the JS version of Verovio in a browser environment, should we prioritize that as "Basic usage" first?

```javascript
const verovio = require('verovio');
const fs = require('fs');

/* Wait for verovio to load */
verovio.module.onRuntimeInitialized = function ()
{
   // create the toolkit instance
   const vrvToolkit = new verovio.toolkit();
   // read the MEI file
	mei = fs.readFileSync('hello.mei');
   // load the MEI data as string into the toolkit
	vrvToolkit.loadData(mei.toString());
   // render the fist page as SVG
	svg = vrvToolkit.renderToSVG(1, {});
   // save the SVG into a file
	fs.writeFileSync('hello.svg', svg);
}
```

## Usage with ESM

> Since version 3.11.0 there is an ESM compatible version of the npm package with a modularized build of the Verovio module, that needs a slightly different setup to wait for the asynchronous Emscripten module to be ready for usage which is now [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) based instead of using the `onRuntimeInitialized` callback function.

Reviewer: Should probably be consistent with capitalizing `npm` vs `NPM`.

Author: Thanks. *npm* should be lower case everywhere. I made them italic everywhere to make it clear that this is expression.

> Since version 3.11.0 there is an ESM compatible version of the npm package with a modularized build of the Verovio module, that needs a slightly different setup to wait for the asynchronous Emscripten module to be ready for usage which is now [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) based instead of using the `onRuntimeInitialized` callback function.

Reviewer: ```suggestion
Since version 3.11.0 there is an ESM compatible version of the npm package with a modularized build of the Verovio module. This is because we need to wait for the asynchronous module to be ready for usage, and this is now [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) based instead of using the `onRuntimeInitialized` callback function.
```

>  Use `.mjs` as file extension when using this directly in Node.js or set `"type": "module"` in your `package.json`.

Reviewer: I'm not sure I really understand what this means... where would I use `.mjs`? It's not in any of the examples...

Author: I hope my comment from before will explain this. You can force ESM usage in Node.js if you use `.mjs` as a file extension. Then you can use the modern import syntax. Also have a loook here: https://github.com/WolfgangDrescher/verovio-npm-examples/tree/master/node-esm

Reviewer: Yes, it helps *me*, but since this is going into the book it will need to help others too, and they probably will not know about your comment. :-D

Reviewer 2: @WolfgangDrescher please integrate the explanations you gave here as comments into the book. Thanks
